### PR TITLE
Using async to properly handle exceptions when multiple threads are run

### PR DESCRIPTION
**Pull request by qrilka**

### CODE REVIEW

**Comment by qrilka** [comment]
@tmcgilchrist @markhibberd is something required for this to be merged?
We were bitten by this bug in production

File: riak.cabal [hunk 1]
```diff
@@ -100,6 +100,7 @@ library
 
   build-depends:
                 aeson                         >= 0.8      && < 1.2,
+                async,
```
Comment (tmcgilchrist) [inline riak.cabal]: Would you mind putting some reasonable bounds on this dependency?

**Comment by tmcgilchrist** [comment]
Do you have some investigation notes from the bug you encountered?

If this change fixed it I will merge but I'm still not convinced the code is doing the right thing under exceptions. We've been bitten badly with async exceptions lately and I'm suspicious of everything :-)

**Comment by qrilka** [comment]
@tmcgilchrist that was 3 months ago and unfortunately I haven't done detailed notes regarding this bug.
But as far as I remember the main problem was that our data processing code reading Riak sometimes was just stuck without giving any logs or something. This patch fixed the situation.
The problem in our code was cause by `getMany` function with was calling `pipe` under the hood.
And in unpatched `pipe` any exception in `replicateM_ numReqs $ writeChan ch =<< receive conn` will just kill sending thread leaving the main one waiting forever for results in channel `ch`.
The patch adds more safety as `asyncBoth` re-throws child thread exception if it happens.

File: riak.cabal (hunk 1 above)
Comment (qrilka) [inline riak.cabal]: amended the commit with `>= 2.0.0.0`

**Comment by tmcgilchrist** [comment]
Thanks for the details, I'll push out a new version once the builds are all good.

**Comment by tmcgilchrist** [comment]
Merged here https://github.com/markhibberd/riak-haskell-client/pull/77

**Pull request closed without merging**